fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/OnPredicateMoveAroundRule.java

```diff
@@ -0,0 +1,273 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.starrocks.sql.optimizer.rule.transformation;
+
+import com.google.common.collect.Lists;
+import com.google.common.collect.Range;
+import com.google.common.collect.Sets;
+import com.starrocks.analysis.BinaryType;
+import com.starrocks.sql.optimizer.OptExpression;
+import com.starrocks.sql.optimizer.OptimizerContext;
+import com.starrocks.sql.optimizer.Utils;
+import com.starrocks.sql.optimizer.base.ColumnRefSet;
+import com.starrocks.sql.optimizer.operator.OperatorType;
+import com.starrocks.sql.optimizer.operator.logical.LogicalFilterOperator;
+import com.starrocks.sql.optimizer.operator.logical.LogicalJoinOperator;
+import com.starrocks.sql.optimizer.operator.pattern.Pattern;
+import com.starrocks.sql.optimizer.operator.scalar.BinaryPredicateOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ConstantOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperator;
+import com.starrocks.sql.optimizer.property.RangeExtractor;
+import com.starrocks.sql.optimizer.property.ReplaceShuttle;
+import com.starrocks.sql.optimizer.property.ValueProperty;
+import com.starrocks.sql.optimizer.property.ValuePropertyDeriver;
+import com.starrocks.sql.optimizer.rule.RuleType;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class OnPredicateMoveAroundRule extends TransformationRule {
+
+    public static final OnPredicateMoveAroundRule INSTANCE = new OnPredicateMoveAroundRule(RuleType.TF_PREDICATE_PROPAGATE,
+            Pattern.create(OperatorType.LOGICAL_JOIN).
+                    addChildren(Pattern.create(OperatorType.PATTERN_LEAF), Pattern.create(OperatorType.PATTERN_LEAF)));
+
+    private OnPredicateMoveAroundRule(RuleType type, Pattern pattern) {
+        super(type, pattern);
+    }
+
+    @Override
+    public boolean check(OptExpression input, OptimizerContext context) {
+        if (!context.getSessionVariable().isEnablePredicateMoveAround()) {
+            return false;
+        }
+        LogicalJoinOperator joinOperator = input.getOp().cast();
+        if (joinOperator.getJoinType().isFullOuterJoin() || joinOperator.getJoinType().isCrossJoin()) {
```

Comment:
What if semi/anti join?

Reply:
Let me add it.

Reply:
semi join is added. anti join the derived predicate may helps little, so I skip it.